packages/tezos-accuser-006-PsCARTHA/tezos-accuser-006-PsCARTHA.7.0~rc1/opam

```diff
@@ -0,0 +1,27 @@
+opam-version: "2.0"
+maintainer: "contact@tezos.com"
+authors: [ "Tezos devteam" ]
+homepage: "https://www.tezos.com/"
+bug-reports: "https://gitlab.com/tezos/tezos/issues"
+dev-repo: "git+https://gitlab.com/tezos/tezos.git"
+license: "MIT"
+depends: [
+  "tezos-tooling" { with-test & = version }
+  "dune" { >= "1.11" }
+  "tezos-baking-006-PsCARTHA-commands" { = version }
+  "tezos-client-base-unix" { = version }
+]
+build: [
+  ["dune" "build" "-p" name "-j" jobs]
+  ["mv" "src/proto_006_PsCARTHA/bin_accuser/%{name}%.install" "./"]
```

Comment:
It would be good if you could get rid of this line in the packages. In normal operation, `dune build -p` will build only the install file for the package you have requested.  Do you need this explicit install override for some other reason?  We may be able to just fix that directly in dune if there's an underlying limitation.

Reply:
Very good question about general organization of the repository. I always wanted to discuss it more broadly with the community without daring :-)

These `mv` are necessary because opam files are not at the root of the repository and dune generates `.install` files next to their corresponding opam file.

In addition of requiring these ugly `mv`, I think (but I'm not sure I wanted to try) that not having the opam files at the root of the repository breaks `opam pin add --dev` which is also unfortunate.

On the contrary, as you see the tezos repository defines 80+ opam packages so having 80+ more files at the root would quite pollute it . Moreover, some part of the code ("economic protocols" in our terminology corresponding to the folders `src/proto_*`) are supposed to be (un)pluggable (self contained) plugins, if they are there, software is able to understand/follow the corresponding protocol, if they aren't, you have valid binaries but that don't know about these specific rulesets (but can ask them to their "peers"/compile them/dynamically load them on necessity... an other story). If the opam files corresponding to the libraries of the protocol "foobar" has now to be at the root of the repo, adding/removing protocol "foobar" is not as simple as adding/removing folder `src/proto_foobar` anymore and that's annoying.

So I'm confused about what direction to follow...

Reply:
Hi @pirbo, glad you were able to release tezos here :-)

The next version of opam should add better support for non-toplevel `.opam`. See: https://github.com/ocaml/opam/pull/3499 Basically, it allow to add a field `subpath` in the package description, meaning: this package only requires this fragment of the `.tar.gz` (or the repository) in order to build. So `opam` only extracts (or fetch) the required `subpath` and build directly in the subdirectory. So, from the dune perspective, both`.opam` and `.install` are at toplevel.

And, to fix the daily workflow, it adds `opam pin add --dev --rec`: il will lookup for `.opam` in sub-directories.

I have no idea whenever the next (feature-)release of opam might happen but, meanwhile, maybe this ugly `mv` is acceptable. And, maybe, you are the perfect beta-tester for this new feature :-)

Reply:
Fantastic! 
To hear from you first and foremost but also this feature looks to be exactly what we want indeed. No surprise given its original author. :-D
I agree that the `mv` are acceptable in the meantime.

Reply:
I'm not quite convinced that the behaviour of dune is quite correct here, though. If the `dune build -p` works, then the semantics of `.install` files are that they should be generated at the root of the project to be useful.  If you could open up an issue over at https://github.com/ocaml/dune/issues about this, we can take a look at supporting this properly.

We'll be kicking off the opam 2.1 release process in the next weeks, to answer @hnrgrgr's question above.  But it would be good to find a solution that'll work with opam 2.0.x as well, since it is widely packaged on OS distributions.